funsor/einsum/numpy_map.py

```diff
@@ -0,0 +1,38 @@
+# Copyright (c) 2017-2019 Uber Technologies, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+import operator
+
+from functools import reduce
+
+import funsor.ops as ops
+from funsor.einsum.util import Tensordot, broadcast_all, rename_equation
+
+
+def einsum(equation, *operands):
+    """
+    Forward-max-sum backward-argmax implementation of einsum.
+    This assumes all operands have a ``._pyro_dims`` attribute set.
+    """
+    # TODO: the implementation skips "backward" logic
+    equation = rename_equation(equation, *operands)
+    inputs, output = equation.split('->')
+
+    contract_dims = ''.join(sorted(set().union(*(x._pyro_dims for x in operands)) - set(output)))
+    dims = output + contract_dims
+    result = reduce(operator.add, broadcast_all(*operands, dims=dims))
```

Comment:
Is this the only place where `_pyro_dims` are actually used and not just passed along?

Reply:
This and `rename_equation`, `broadcast_all` utility require `pyro_dims`. I am trying to remove `pyro_dims`. If things pass, then I think we can simplify much code here.